src/OpenTelemetry/Internal/InterlockedHelper.cs

```diff
@@ -18,29 +18,28 @@ public static void Add(ref double location, double value)
         var returnedValue = Interlocked.CompareExchange(ref location, currentValue + value, currentValue);
         if (returnedValue != currentValue)
         {
-            AddRare(ref location, value, returnedValue);
+            AddRare(ref location, value);
         }
     }
 
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public static double Read(ref double location)
-        => Interlocked.CompareExchange(ref location, double.NaN, double.NaN);
+        => Interlocked.CompareExchange(ref location, 0, 0);
 
     [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void AddRare(ref double location, double value, double currentValue)
+    private static void AddRare(ref double location, double value)
     {
         var sw = default(SpinWait);
         while (true)
         {
             sw.SpinOnce();
 
+            double currentValue = Volatile.Read(ref location);
             var returnedValue = Interlocked.CompareExchange(ref location, currentValue + value, currentValue);
```

Comment:
Doesn't just refe.tching the new value break the compare exchange pattern?

If that is not needed, you can just overwrite what is there, without looking at the current value.

Reply:
Not sure if I understood the question correctly, but the new value is calculated based on the current value. Previously it use `currentValue` from previous failed compare-exchange, but after spin waiting, which doesn't make much sense since the value is quite old by that point (and thus likely to fail again).

Reply:
I haven't seen spin waits used with compare exchange in simple flows like this where it's guaranteed that forward progress is made even under heavy contention. So maybe should remove the spin wait and then it would make sense to use the compare exchange result for current value.

Reply:
I removed the spin wait since just doing compare exchange in a loop is more efficient and produces smaller machine code.